pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
 	}
 
 	switch operation {
 	case "start":
-		var pid uint32 = 0
-
 		if trace.Spec.Filter == nil {
 			gadgets.SetStatusError(trace, "Missing filter")
 			return
 		}
-		if trace.Spec.Filter.Namespace == "" {
-			gadgets.SetStatusError(trace, "Invalid filter: missing namespace")
-			return
-		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
-			return
-		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
-			return
-		}
 		if trace.Spec.Filter.ContainerName != "" {
-			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
+			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!",
+				trace.Spec.Gadget)
 		}
 
-		pidsInPod := t.resolver.LookupPIDByPod(trace.Spec.Filter.Namespace, trace.Spec.Filter.Podname)
-
-		// All containers inside a pod share the same network namespace
-		// thus we can just take the first valid pid we found
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
 		}
 
-		log.Infof("Gadget %s: Using PID %d to retrieve network namespaces of Podname %q in Namespace %q",
-			trace.Spec.Gadget, pid, trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace)
-
-		if pid == 0 {
-			gadgets.SetStatusError(trace, fmt.Sprintf("Couldn't find a valid PID for Podname %q in Namespace %q",
-				trace.Spec.Filter.Podname, trace.Spec.Filter.Namespace))
+		output, err := start(trace.Spec.Gadget, trace.Spec.Node, filteredContainers)
+		if err != nil {
+			gadgets.SetStatusError(trace, err.Error())
 			return
 		}
 
-		t.Start(trace, pid)
+		trace.Status.OperationError = ""
+		trace.Status.Output = output
+		trace.Status.State = "Completed"
 	default:
 		gadgets.SetStatusError(trace, fmt.Sprintf("Unknown operation %q", operation))
 	}
 }
 
-func (t *Trace) Start(trace *gadgetv1alpha1.Trace, pid uint32) {
-	output, err := tracer.RunCollector(
-		pid,
-		trace.Spec.Filter.Podname,
-		trace.Spec.Filter.Namespace,
-		trace.Spec.Node,
-	)
+func start(gadgetName, node string, filteredContainers []pb.ContainerDefinition) (string, error) {
+	allSockets := []socketcollectortypes.Event{}
+
+	// Given that the socket-collector tracer works per network namespace and
+	// all the containers inside a pod/namespace share the network namespace,
+	// we only need to run the tracer with one valid PID per pod/namespace
+	visitedPods := make(map[string]bool)
+
+	for _, container := range filteredContainers {
+		key := container.Podname + "/" + container.Namespace
```

Comment:
Nit: usually the namespace is appended before.

Reply:
Done. I like the code to be uniform 🤓